lib/bolt/bolt_option_parser.rb

```diff
@@ -136,53 +146,53 @@ def initialize(options)
-                        'For Windows targets, specify the winrm:// protocol if it has not be configured',
-                        'For SSH, port defaults to `22`',
-                        'For WinRM, port defaults to `5985` or `5986` based on the --[no-]ssl setting') do |targets|
+      define('-t', '--targets TARGETS',
+             'Identifies the targets of command.',
+             'Enter a comma-separated list of target URIs or group names.',
+             "Or read a target list from an input file '@<file>' or stdin '-'.",
+             'Example: --targets localhost,node_group,ssh://nix.com:23,winrm://windows.puppet.com',
+             'URI format is [protocol://]host[:port]',
+             "SSH is the default protocol; may be #{TRANSPORTS.keys.join(', ')}",
+             'For Windows targets, specify the winrm:// protocol if it has not be configured',
+             'For SSH, port defaults to `22`',
+             'For WinRM, port defaults to `5985` or `5986` based on the --[no-]ssl setting') do |targets|
         @options[:targets] ||= []
         @options[:targets] << get_arg_input(targets)
       end.extend(SwitchHider)
-      @query = define('-q', '--query QUERY', 'Query PuppetDB to determine the targets') do |query|
+      define('-q', '--query QUERY', 'Query PuppetDB to determine the targets') do |query|
         @options[:query] = query
       end.extend(SwitchHider)
-      @rerun = define('--rerun FILTER', 'Retry on nodes from the last run',
-                      "'all' all nodes that were part of the last run.",
-                      "'failure' nodes that failed in the last run.",
-                      "'success' nodes that succeeded in the last run.") do |rerun|
+      define('--rerun FILTER', 'Retry on nodes from the last run',
+             "'all' all nodes that were part of the last run.",
+             "'failure' nodes that failed in the last run.",
+             "'success' nodes that succeeded in the last run.") do |rerun|
         @options[:rerun] = rerun
       end.extend(SwitchHider)
       define('--noop', 'Execute a task that supports it in noop mode') do |_|
         @options[:noop] = true
-      end
+      end.extend(SwitchHider)
       define('--description DESCRIPTION',
              'Description to use for the job') do |description|
         @options[:description] = description
-      end
+      end.extend(SwitchHider)
       define('--params PARAMETERS',
              "Parameters to a task or plan as json, a json file '@<file>', or on stdin '-'") do |params|
         @options[:task_options] = parse_params(params)
-      end
-      @execute = define('-e', '--execute CODE',
-                        "Puppet manifest code to apply to the targets") do |code|
+      end.extend(SwitchHider)
+      define('-e', '--execute CODE',
+             "Puppet manifest code to apply to the targets") do |code|
         @options[:code] = code
       end.extend(SwitchHider)
 
       separator 'Authentication:'
```

Comment:
One side affect of hiding options is that the separators are left behind. This doesn't happen with most of the commands, but for one like 'puppetfile' you'll have several separators with no options listed under them.

Should these be removed or should some logic be added to turn them off when all associated options are hidden?

Reply:
I think we can get rid of the `SwitchHider` class entirely. It's not really what we want, since it only causes unsupported flags to not be _showed_, but they will still be _accepted_ and then we rely on later validation to ensure we fail if they're not welcome.

Because we now have an explicit filter step, we could potentially just _delete_ the excluded flags rather than extend them all with the `SwitchHider`. In that case, we could keep track of the `top` stack and, if we've deleted every flag up to a specific separator, also delete the separator.